src/ccsr/src/client.rs

```diff
@@ -162,6 +162,29 @@ impl Client {
         let _res: Vec<i32> = send_request(req).await?;
         Ok(())
     }
+
+    /// Gets the latest version of the first subject found associated with the scheme with
+    /// the given id, as well as all other subjects referenced by that subject (recursively).
+    ///
+    /// The dependencies are returned in alphabetical order by subject name.
+    pub async fn get_subject_and_references_by_id(
+        &self,
+        id: i32,
+    ) -> Result<(Subject, Vec<Subject>), GetBySubjectError> {
+        for name in self.list_subjects().await.map_err(|le| match le {
+            ListError::Transport(t) => GetBySubjectError::Transport(t),
+            ListError::Server { code, message } => GetBySubjectError::Server { code, message },
+        })? {
+            let subject = self.get_subject(&name).await?;
+
+            // We just assume the first one we find is correct
+            if subject.schema.id == id {
+                return self.get_subject_and_references(&subject.name).await;
+            }
+        }
+
+        Err(GetBySubjectError::SubjectNotFound)
+    }
```

Comment:
Eek! This is going to be very slow when there are a lot of subjects. This should look more like `get_schema_by_id` where it uses the dedicated API for looking a schema up by ID. You'll probably need to factor out the "recursively fetch references" code into its own helper function.

Reply:
That api doesn't exist? https://docs.confluent.io/platform/current/schema-registry/develop/using.html#list-all-subjects-associated-with-a-given-id  (and the tip here: https://docs.confluent.io/platform/current/schema-registry/develop/api.html#subjects)

As far as I can tell, no api's give you subjects from schema id's, which you need to get references (which we need, to compile), all the schema fetching code don't return references? (that said, the docs are definitely problematic, they don't mention the references we use in `GetBySubjectResponse`, ill play around with the api and see if the schema lookup api gives and more info)

Definitely very supoptimal and I'd be curious to hear any better solutions?

Reply:
omggg its actually IS undocumented that there are references in that response! Will definitely refactor and fix it up!

Reply:
actually, https://docs.confluent.io/platform/current/schema-registry/develop/api.html#get--schemas-ids-int-%20id-versions is even simpler!

however, its still subtle, and I will a comment explaining the correctness for you to review

Reply:
I think this is viable, but you can't lose the version that comes back or you risk getting the wrong schema! `get_subject_and_references` will need to learn to take a `version: Option<i32>` parameter that defaults to `latest` if unspecified, and then the version from the response needs to get passed through.

Reply:
Oh i see, that makes sense, and is an easy fix!

Reply:
I refactored it in a different way (separate, non-pub version that requires a version)